File: backend/mr/src/main/java/io/featurehub/mr/resources/PersonResource.java
package io.featurehub.mr.resources;

import cd.connect.app.config.ConfigKey;
import cd.connect.app.config.DeclaredConfigResolver;
import io.featurehub.db.api.FillOpts;
import io.featurehub.db.api.GroupApi;
import io.featurehub.db.api.OptimisticLockingException;
import io.featurehub.db.api.Opts;
import io.featurehub.db.api.PersonApi;
import io.featurehub.db.services.Conversions;
import io.featurehub.mr.api.PersonServiceDelegate;
import io.featurehub.mr.auth.AuthManagerService;
import io.featurehub.mr.model.CreatePersonDetails;
import io.featurehub.mr.model.OutstandingRegistration;
import io.featurehub.mr.model.Person;
import io.featurehub.mr.model.RegistrationUrl;
import io.featurehub.mr.model.SearchPersonResult;
import jakarta.inject.Inject;
import jakarta.ws.rs.BadRequestException;
import jakarta.ws.rs.ForbiddenException;
import jakarta.ws.rs.NotFoundException;
import jakarta.ws.rs.WebApplicationException;
import jakarta.ws.rs.core.Response;
import jakarta.ws.rs.core.SecurityContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Optional;
import java.util.UUID;
import java.util.stream.Collectors;

public class PersonResource implements PersonServiceDelegate {
  private static final Logger log = LoggerFactory.getLogger(PersonResource.class);
  private final PersonApi personApi;
  private final GroupApi groupApi;
  private final AuthManagerService authManager;

  @ConfigKey("register.url")
  private String registrationUrl = "http://localhost:register-url?token=%s";

  @Inject
  public PersonResource(PersonApi personApi, GroupApi groupApi, AuthManagerService authManager) {
    this.personApi = personApi;
    this.groupApi = groupApi;
    this.authManager = authManager;

    DeclaredConfigResolver.resolve(this);
  }

  @Override
  public RegistrationUrl createPerson(CreatePersonDetails createPersonDetails, CreatePersonHolder holder,
                                      SecurityContext securityContext) {
    Person currentUser = authManager.from(securityContext);

    if (authManager.isAnyAdmin(currentUser.getId().getId())) {
      //create new user in the db
      try {
        PersonApi.PersonToken person = personApi.create(createPersonDetails.getEmail(),
          createPersonDetails.getName(), currentUser.getId().getId());

        if (person == null || person.id == null) {
          throw new BadRequestException();
        }

        if (createPersonDetails.getGroupIds() != null) {
          //add user to the groups
          Optional.of(createPersonDetails.getGroupIds()).ifPresent(list -> list.forEach(id -> {
            groupApi.addPersonToGroup(id, person.id, Opts.empty());
          }));
        }

        //return registration url
        RegistrationUrl regUrl = new RegistrationUrl();
        // hard code the return value, it will be ignored by the client from now on
        regUrl.setRegistrationUrl(String.format(registrationUrl, person.token));
        regUrl.setToken(person.token);
        return regUrl;
      } catch (PersonApi.DuplicatePersonException e) {
        throw new WebApplicationException(Response.status(Response.Status.CONFLICT).build());
      }
    }

    throw new ForbiddenException("Not admin");
  }

  private Opts peopleOpts(Boolean includeAcls, Boolean includeGroups) {
    return new Opts()
      .add(FillOpts.Groups, includeGroups)
      .add(FillOpts.Acls, includeAcls);
  }


  private Person getPerson(String id, Boolean includeAcls, Boolean includeGroups, SecurityContext securityContext) {
    Person currentUser = authManager.from(securityContext);


    UUID personId;
    if ("self".equals(id)) {
      personId = currentUser.getId().getId();
      log.debug("User requested their own details: {}", id);
    } else {
      personId = Conversions.checkUuid(id);
    }

    if (currentUser.getId().getId().equals(personId) || authManager.isAnyAdmin(currentUser.getId().getId())) {
      Person person;

      if (personId == null) {
        person = personApi.get(id, peopleOpts(includeAcls, includeGroups));
      } else {
        person = personApi.get(personId, peopleOpts(includeAcls, includeGroups));
      }

      if (person == null) {
        throw new NotFoundException();
      }

      return person;
    }

    throw new ForbiddenException("You are not allowed the details of this person");
  }

  @Override
  public Boolean deletePerson(String id, DeletePersonHolder holder, SecurityContext securityContext) {
    if (authManager.isOrgAdmin(authManager.from(securityContext))) {
      Person p = getPerson(id, false, false, securityContext);

      if (p.getEmail() != null) {
        return personApi.delete(p.getEmail());
      }
    }

    throw new ForbiddenException("No permission");
  }

  @Override
  public SearchPersonResult findPeople(FindPeopleHolder holder, SecurityContext securityContext) {
    Person currentUser = authManager.from(securityContext);

    if (authManager.isAnyAdmin(currentUser.getId().getId())) {

      int start = holder.startAt == null ? 0 : holder.startAt;
      int page = holder.pageSize == null ? 20 : holder.pageSize;

      PersonApi.PersonPagination pp = personApi.search(holder.filter, holder.order, start, page,
        new Opts().add(FillOpts.Groups, holder.includeGroups).add(FillOpts.PersonLastLoggedIn, holder.includeLastLoggedIn));

      return new SearchPersonResult()
        .people(pp.people)
        .outstandingRegistrations(pp.personsWithOutstandingTokens.stream().map(
          pt -> new OutstandingRegistration().id(pt.id).token(pt.token).expired(pp.personIdsWithExpiredTokens.contains(pt.id))
        ).collect(Collectors.toList()))
        .max(pp.max);
    }

    throw new ForbiddenException("Not admin");
  }

  @Override
  public Person getPerson(String id, GetPersonHolder holder, SecurityContext securityContext) {
    return getPerson(id, holder.includeAcls, holder.includeGroups, securityContext);
  }

  @Override
  public Person updatePerson(String id, Person person, UpdatePersonHolder holder, SecurityContext securityContext) {
    Person from = authManager.from(securityContext);

    if (authManager.isAnyAdmin(from)) {
      Person updatedPerson = null;
      try {
        updatedPerson = personApi.update(Conversions.checkUuid(id), person,
          peopleOpts(holder.includeAcls, holder.includeGroups),
          from.getId().getId());
      } catch (OptimisticLockingException e) {
        throw new WebApplicationException(422);
      } catch (IllegalArgumentException iae) {
        throw new BadRequestException();
      }

      if (updatedPerson == null) {
        throw new NotFoundException();
      }

      return updatedPerson;
    }

    throw new ForbiddenException("Not authorised");
  }
}

File: backend/mr/src/main/kotlin/io/featurehub/mr/resources/PersonResource.kt
package io.featurehub.mr.resources

import cd.connect.app.config.ConfigKey
import cd.connect.app.config.DeclaredConfigResolver
import io.featurehub.db.api.*
import io.featurehub.db.services.Conversions
import io.featurehub.mr.api.PersonServiceDelegate
import io.featurehub.mr.auth.AuthManagerService
import io.featurehub.mr.model.*
import jakarta.inject.Inject
import jakarta.ws.rs.BadRequestException
import jakarta.ws.rs.ForbiddenException
import jakarta.ws.rs.NotFoundException
import jakarta.ws.rs.WebApplicationException
import jakarta.ws.rs.core.Response
import jakarta.ws.rs.core.SecurityContext
import org.slf4j.LoggerFactory
import java.util.*
import java.util.Set
import java.util.stream.Collectors

class PersonResource @Inject constructor(
  private val personApi: PersonApi,
  private val groupApi: GroupApi,
  private val authManager: AuthManagerService
) : PersonServiceDelegate {

  @ConfigKey("register.url")
  private var registrationUrl: String? = "http://localhost:register-url?token=%s"

  init {
    DeclaredConfigResolver.resolve(this)
  }

  override fun createPerson(
    createPersonDetails: CreatePersonDetails, holder: PersonServiceDelegate.CreatePersonHolder,
    securityContext: SecurityContext
  ): RegistrationUrl {
    val currentUser = authManager.from(securityContext)

    if (!authManager.isAnyAdmin(currentUser.id!!.id)) {
      throw ForbiddenException("Not allowed")
    }

    if (createPersonDetails.personType == PersonType.SERVICEACCOUNT) {
      return createServiceAccount(createPersonDetails, currentUser)
    }

    if (createPersonDetails.email == null) {
      throw BadRequestException("Email is a required parameter for creating a person.")
    }

    //create new user in the db
    return try {
      val person = personApi.create(
        createPersonDetails.email!!,
        createPersonDetails.name, currentUser.id!!.id
      )
      if (person?.id == null) {
        throw BadRequestException()
      }

      createPersonDetails.groupIds?.forEach { id ->
        groupApi.addPersonToGroup(id, person.id, Opts.empty())
      }

      //return registration url
      RegistrationUrl()
        .personId(person.id)
        .registrationUrl(
          String.format(
            registrationUrl!!,
            person.token
          )
        ) // hard code the return value, it will be ignored by the client from now on
        .token(person.token)
    } catch (e: PersonApi.DuplicatePersonException) {
      throw WebApplicationException(Response.status(Response.Status.CONFLICT).build())
    }
  }

  private fun createServiceAccount(
    createPersonDetails: CreatePersonDetails,
    currentUser: Person
  ): RegistrationUrl {
    if (createPersonDetails.name == null) {
      throw BadRequestException("Name is required parameter")
    }

    val (person, token) = personApi.createServicePerson(
      createPersonDetails.name!!,
      currentUser.id!!.id
    ) ?: throw BadRequestException()
    val servicePersonId = person.id!!.id

    createPersonDetails.groupIds?.forEach { id ->
      groupApi.addPersonToGroup(id, servicePersonId, Opts.empty())
    }

    // return registration url
    return RegistrationUrl()
      .personId(servicePersonId)
      .token(token)
  }

  private fun peopleOpts(includeAcls: Boolean?, includeGroups: Boolean?): Opts {
    return Opts()
      .add(FillOpts.Groups, includeGroups)
      .add(FillOpts.Acls, includeAcls)
  }

  private fun getPerson(
    id: String,
    includeAcls: Boolean?,
    includeGroups: Boolean?,
    securityContext: SecurityContext
  ): Person {
    val currentUser = authManager.from(securityContext)

    val personId: UUID?
    if ("self" == id) {
      personId = currentUser.id!!.id
      log.debug("User requested their own details: {}", id)
    } else {
      personId = Conversions.checkUuid(id)
    }

    if (currentUser.id!!.id == personId || authManager.isAnyAdmin(currentUser.id!!.id)) {
      val person = if (personId == null) {
        personApi[id, peopleOpts(includeAcls, includeGroups)]
      } else {
        personApi[personId, peopleOpts(includeAcls, includeGroups)]
      }
      if (person == null) {
        throw NotFoundException()
      }
      return person
    }
    throw ForbiddenException("You are not allowed the details of this person")
  }

  override fun deletePerson(
    id: UUID,
    holder: PersonServiceDelegate.DeletePersonHolder,
    securityContext: SecurityContext
  ): Boolean {
    if (authManager.isOrgAdmin(authManager.from(securityContext))) {
      val p = getPerson(id.toString(), false, false, securityContext)
      if (p.email != null) {
        return personApi.delete(p.email!!, holder.includeGroups == true)
      }
    }
    throw ForbiddenException("No permission")
  }

  override fun findPeople(
    holder: PersonServiceDelegate.FindPeopleHolder,
    securityContext: SecurityContext
  ): SearchPersonResult {
    val currentUser = authManager.from(securityContext)
    if (!authManager.isAnyAdmin(currentUser.id!!.id)) throw ForbiddenException("Not Admin")

    val start = if (holder.startAt == null) 0 else holder.startAt
    val page = if (holder.pageSize == null) 20 else holder.pageSize

    val pp = personApi.search(
      holder.filter, holder.order, start, page,
      if (holder.personTypes == null) Set.of(PersonType.PERSON) else HashSet(holder.personTypes),
      holder.sortBy,
      Opts().add(FillOpts.Groups, holder.includeGroups)
        .add(FillOpts.CountGroups, holder.countGroups)
        .add(FillOpts.Archived, holder.includeDeactivated)
        .add(FillOpts.PersonLastLoggedIn, holder.includeLastLoggedIn)
    )

    val persons = if (holder.countGroups == true)
      SearchPersonResult().summarisedPeople(pp.searchPeople)
    else
      SearchPersonResult().people(pp.people)

    return persons
      .outstandingRegistrations(
        pp.personsWithOutstandingTokens.stream().map { pt: PersonApi.PersonToken ->
          OutstandingRegistration().id(pt.id).token(pt.token)
            .expired(pp.personIdsWithExpiredTokens.contains(pt.id))
        }
          .collect(Collectors.toList()))
      .max(pp.max)
  }

  override fun getPerson(
    id: String,
    holder: PersonServiceDelegate.GetPersonHolder,
    securityContext: SecurityContext
  ): Person {
    return getPerson(id, holder.includeAcls, holder.includeGroups, securityContext)
  }

  override fun resetSecurityToken(id: UUID, securityContext: SecurityContext): AdminServiceResetTokenResponse {
    val currentUser = authManager.from(securityContext)
    if (authManager.isAnyAdmin(currentUser.id!!.id)) {
      val updatedServicePerson = personApi.resetServicePersonToken(id)
      if (updatedServicePerson != null) {
        return AdminServiceResetTokenResponse().token(updatedServicePerson.token)
      }
      throw NotFoundException()
    }
    throw ForbiddenException()
  }

  override fun updatePerson(
    id: UUID,
    person: Person,
    holder: PersonServiceDelegate.UpdatePersonHolder,
    securityContext: SecurityContext
  ): Person {
    val from = authManager.from(securityContext)
    if (authManager.isAnyAdmin(from)) {
      return try {
        personApi.update(
          id, person,
          peopleOpts(holder.includeAcls, holder.includeGroups),
          from.id!!.id
        )
      } catch (e: OptimisticLockingException) {
        throw WebApplicationException(422)
      } catch (iae: IllegalArgumentException) {
        throw BadRequestException()
      } ?: throw NotFoundException()
    }
    throw ForbiddenException("Not authorised")
  }

  override fun updatePersonV2(
    id: UUID,
    updatePerson: UpdatePerson,
    securityContext: SecurityContext
  ) {
    val from = authManager.from(securityContext)

    if (authManager.isAnyAdmin(from)) {
      try {
        personApi.updateV2(
          id, updatePerson,
          from.id!!.id
        )
      } catch (e: OptimisticLockingException) {
        throw WebApplicationException(422)
      } catch (iae: IllegalArgumentException) {
        throw BadRequestException()
      } ?: throw NotFoundException()

      return
    }

    throw ForbiddenException("Not authorised")
  }

  companion object {
    private val log = LoggerFactory.getLogger(PersonResource::class.java)
  }
}
